components/canvas/canvas_data.rs

```diff
@@ -18,16 +18,175 @@ use std::mem;
 use std::sync::Arc;
 use webrender_api;
 
+/// The canvas data stores a state machine for the current status of
+/// the path data and any relevant transformations that are
+/// applied to it. The Azure drawing API expects the path to be in
+/// userspace. However, when a path is being built but the canvas'
+/// transform changes, we choose to transform the path and perform
+/// further operations to it in device space. When it's time to
+/// draw the path, we convert it back to userspace and draw it
+/// with the correct transform applied.
+enum PathState {
+    /// Path builder in user-space. If a transform has been applied
+    /// but no further path operations have occurred, it is stored
+    /// in the optional field.
+    UserSpacePathBuilder(PathBuilder, Option<Transform2D<AzFloat>>),
+    /// Path builder in device-space.
+    DeviceSpacePathBuilder(PathBuilder),
+    /// Path in user-space. If a transform has been applied but
+    /// but no further path operations have occurred, it is stored
+    /// in the optional field.
+    UserSpacePath(Path, Option<Transform2D<AzFloat>>),
+}
+
+impl PathState {
+    fn is_path(&self) -> bool {
+        match *self {
+            PathState::UserSpacePath(..) => true,
+            PathState::UserSpacePathBuilder(..) |
+            PathState::DeviceSpacePathBuilder(..) => false,
+        }
+    }
+
+    fn path(&self) -> &Path {
+        match *self {
+            PathState::UserSpacePath(ref p, _) => p,
+            PathState::UserSpacePathBuilder(..) |
+            PathState::DeviceSpacePathBuilder(..) => panic!("should have called ensure_path"),
+        }
+    }
+}
+
+/// A wrapper around a stored PathBuilder and an optional transformation that should be
+/// applied to any points to ensure they are in the matching device space.
+struct PathBuilderRef<'a> {
+    builder: &'a PathBuilder,
+    transform: Option<Transform2D<AzFloat>>,
```

Comment:
I wonder if this could just be the identity transform instead of `None`. That would simplify the code below…